crates/bevy_ui/src/render/gradient.wgsl

```diff
@@ -0,0 +1,280 @@
+    // * Negative values indicate that the point is inside the shape.
+    // * Zero values indicate the point is on the edge of the shape.
+    // * Positive values indicate the point is outside the shape.
+
+    // Signed distance from the exterior boundary.
+    let external_distance = sd_rounded_box(in.point, in.size, in.radius);
+
+    // Signed distance from the border's internal edge (the signed distance is negative if the point 
+    // is inside the rect but not on the border).
+    // If the border size is set to zero, this is the same as the external distance.
+    let internal_distance = sd_inset_rounded_box(in.point, in.size, in.radius, in.border);
+
+    // Signed distance from the border (the intersection of the rect with its border).
+    // Points inside the border have negative signed distance. Any point outside the border, whether 
+    // outside the outside edge, or inside the inner edge have positive signed distance.
+    let border_distance = max(external_distance, -internal_distance);
+
+#ifdef ANTI_ALIAS
+    // At external edges with no border, `border_distance` is equal to zero. 
+    // This select statement ensures we only perform anti-aliasing where a non-zero width border 
+    // is present, otherwise an outline about the external boundary would be drawn even without 
+    // a border.
+    let t = select(1.0 - step(0.0, border_distance), antialias(border_distance), external_distance < internal_distance);
+#else
+    let t = 1.0 - step(0.0, border_distance);
+#endif
+
+    // Blend mode ALPHA_BLENDING is used for UI elements, so we don't premultiply alpha here.
+    return vec4(color.rgb, saturate(color.a * t));
+}
+
+fn draw_background(in: VertexOutput, color: vec4<f32>) -> vec4<f32> {
+    // When drawing the background only draw the internal area and not the border.
+    let internal_distance = sd_inset_rounded_box(in.point, in.size, in.radius, in.border);
+
+#ifdef ANTI_ALIAS
+    let t = antialias(internal_distance);
+#else
+    let t = 1.0 - step(0.0, internal_distance);
+#endif
+
+    return vec4(color.rgb, saturate(color.a * t));
+}
+
+@fragment
+fn fragment(in: VertexOutput) -> @location(0) vec4<f32> {
+    var g_distance: f32;
+    if enabled(in.flags, RADIAL) {
+        g_distance = radial_distance(in.point, in.g_start, in.dir.x);
+    } else if enabled(in.flags, CONIC) {
+        g_distance = conic_distance(in.point, in.g_start);
+    } else {
+        g_distance = linear_distance(in.point, in.g_start, in.dir);
+    }
+
+    let gradient_color = interpolate_gradient(
+        g_distance,
+        in.start_color,
+        in.start_len,
+        in.end_color,
+        in.end_len,
+        in.hint,
+        in.flags
+    );
+
+    if enabled(in.flags, BORDER) {
+        return draw(in, gradient_color);
+    } else {
+        return draw_background(in, gradient_color);
+    }
+}
+
+fn srgb_mix(a: vec4<f32>, b: vec4<f32>, t: f32) -> vec4<f32> {
```

Comment:
this should be in a util file (and might already exist somewhere)

Reply:
I wasn't able to find anything. I could move it into `color_operations.wgsl` in `bevy_render` but I'm not sure as this 
implemenation might not be ideal. Maybe the colors should be passed into the shader in srgb space? I want to add support for interpolation in other color spaces eventually so didn't want to overthink things.